lib/logstash/outputs/elasticsearch_java/protocol.rb

```diff
@@ -0,0 +1,248 @@
+                request.routing(args[:_routing]) if args[:_routing]
+              when "create"
+                request = org.elasticsearch.action.index.IndexRequest.new(args[:_index])
+                request.id(args[:_id]) if args[:_id]
+                request.routing(args[:_routing]) if args[:_routing]
+                request.source(source)
+                request.opType("create")
+              when "create_unless_exists"
+                unless args[:_id].nil?
+                  request = org.elasticsearch.action.index.IndexRequest.new(args[:_index])
+                  request.id(args[:_id])
+                  request.routing(args[:_routing]) if args[:_routing]
+                  request.source(source)
+                  request.opType("create")
+                else
+                  raise(LogStash::ConfigurationError, "Specifying action => 'create_unless_exists' without a document '_id' is not supported.")
+                end
+              when "update"
+                unless args[:_id].nil?
+                  request = org.elasticsearch.action.update.UpdateRequest.new(args[:_index], args[:_type], args[:_id])
+                  request.routing(args[:_routing]) if args[:_routing]
+                  request.doc(source)
+                  if @options[:doc_as_upsert]
+                    request.docAsUpsert(true)
+                  else
+                    request.upsert(args[:_upsert]) if args[:_upsert]
+                  end
+                else
+                  raise(LogStash::ConfigurationError, "Specifying action => 'update' without a document '_id' is not supported.")
+                end
+              else
+                raise(LogStash::ConfigurationError, "action => '#{action_name}' is not currently supported.")
+            end # case action
+
+            request.type(args[:_type]) if args[:_type]
+            return request
+          end # def build_request
+
+          def template_exists?(name)
+            return !client.admin().indices().
+              prepareGetTemplates(name).
+              execute().
+              actionGet().
+              getIndexTemplates().
+              isEmpty
+          end # def template_exists?
+
+          def template_put(name, template)
+            response = client.admin.indices.
+              preparePutTemplate(name).
+              setSource(LogStash::Json.dump(template)).
+              execute().
+              actionGet()
+
+            raise "Could not index template!" unless response.isAcknowledged
+          end # template_put
+
+          public(:initialize, :bulk)
+        end # class NodeClient
+
+        class TransportClient < NodeClient
+          private
+          def build_client(options)
+            client = org.elasticsearch.client.transport.TransportClient.
+              builder().
+              settings((settings.build)).
+              build()
+
+            options[:hosts].each do |host|
+              matches = host.match /(.+)(?:.*)/
```

Comment:
I don't think this regexp is right, it matches `.+` (one or more of any character) then `.*` (zero or more of any character). The second group is uncaptured.

Did you mean to capture the port? Below, `matches[2]` doesn't exist because there's no second captured group. Something like `/^(.+?)(?::([0-9]+))?$/` would almost work except not for ipv6 addresses.

```
>> "foo:123".match(/^(.+?)(?::([0-9]+))?$/)
=> #<MatchData "foo:123" 1:"foo" 2:"123">
```